### PR TITLE
[WIP] Variance tracking for DTensor manual APIs

### PR DESCRIPTION
**Pull request by vishal9-team**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #173114
* #173107
* __->__ #169620

**Design Document: [gdoc](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu)**

Use cases below previously required precise Placements to be specified by user via ```in_grad_placements``` parameter in local_map

### Use Case 1: Data Parallel:  [**_test_data_parallel_**]
Inserts all_reduce in backward for Parameters

data_parallel

### Use Case 2: Megatron Tensor Parallel: [**_test_megatron_tp_parallel_**]

Algorithm: https://arxiv.org/pdf/1909.08053

megatron_tp

Inserts all_reduce in backward for Input and Parameters as appropriate

graph_ltensor_megatron_tp

### Use Case 3: FSDP + TP: [**_test_megatron_tp_parallel_**]

TP requires all_reduce in backward for input and FSDP requires reduce_scatter for parameters - both are a correctly added

graph_ltensor_rs

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo

### CODE REVIEW

File: test/distributed/test_ltensor_variance_tracking.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
+# Owner(s): ["oncall: distributed"]
+
+import sys
+
+import torch
+import torch.distributed as dist
+from torch.distributed.device_mesh import DeviceMesh
+from torch.distributed.tensor.ltensor.ltensor import LTensor
```
Comment (wconstab) [inline test/distributed/test_ltensor_variance_tracking.py]: LTensor.. haha
so, i actually wonder if it is too late to rename `LocalTensor` to something else and use that name here?

cc @dzmitry-huba

File: test/distributed/test_variance_tracking_ltensor.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    def test_invariant_union(self):
+        """Invariant + Invariant = Invariant."""
+        mesh = DeviceMesh(
+            DEVICE,
+            torch.arange(self.world_size).reshape(2, 2),
+            mesh_dim_names=("dp", "tp"),
+        )
+
+        x = LTensor(torch.randn(4, 8), set(), mesh)
```
Comment (wconstab) [inline test/distributed/test_variance_tracking_ltensor.py:44]: just a 'first impression' thing- this isn't very intuitive to me.  

* why are we passing an empty set? should this argument just be optional and we don't pass anything for this case?

File: test/distributed/test_ltensor_variance_tracking.py (hunk 1 above)
Comment (dzmitry-huba) [reply test/distributed/test_ltensor_variance_tracking.py]: If anything, I'm to blame for the LTensor name :) Vishal asked me if LocalTensor can be renamed and I suggested to use LTensor name (L for Local similar D for distributed). If people are feeling strongly about renaming LocalTensor, let's discuss it before taking renaming (there are quite a few places)

File: test/distributed/test_variance_tracking_ltensor.py (hunk 2 above)
Comment (vishal9-team) [reply test/distributed/test_variance_tracking_ltensor.py:44]: I agree - currently I'm still looking to get the correctness verified with parallelisms and then will clean up the API stuff if it remains unchanged

File: torch/distributed/tensor/_ltensor.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    _variant_axes: set[str]
+    _mesh: DeviceMesh
+
+    def __new__(
+        cls,
+        local_tensor: torch.Tensor,
+        variant_axes: set[str],
+        mesh: DeviceMesh,
+    ):
+        r = local_tensor.as_subclass(cls)
```
Comment (ezyang) [inline torch/distributed/tensor/_ltensor.py:152]: good, don't need a wrapper subclass here

File: torch/distributed/tensor/_ltensor.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    @property
+    def variant_axes(self) -> set[str]:
+        return self._variant_axes
+
+    @property
+    def mesh(self) -> DeviceMesh:
+        return self._mesh
+
+    @classmethod
+    def __torch_function__(cls, func, types, args=(), kwargs=None):
```
Comment (ezyang) [inline torch/distributed/tensor/_ltensor.py:186]: How come a torch function?

File: torch/distributed/tensor/experimental/_func_map.py [hunk 5]
```diff
@@ -31,6 +31,7 @@ def local_map(
     device_mesh: Optional[DeviceMesh] = None,
     *,
     redistribute_inputs: bool = False,
+    track_variant_axes: bool = False,
```
Comment (ezyang) [inline torch/distributed/tensor/experimental/_func_map.py]: No `check_vma` here? lol

File: torch/distributed/tensor/_varying_collectives.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
+# mypy: allow-untyped-defs
+"""Functional collectives with variant/invariant tensor/gradient semantics."""
```
Comment (albanD) [inline torch/distributed/tensor/_varying_collectives.py:2]: Could you document in much more detail what each of these ops are doing?

File: torch/distributed/tensor/experimental/_func_map.py (hunk 5 above)
Comment (vishal9-team) [reply torch/distributed/tensor/experimental/_func_map.py]: It may still turn up if we want to enforce strict variant matching etc. :D

File: torch/distributed/tensor/_ltensor.py (hunk 4 above)
Comment (vishal9-team) [reply torch/distributed/tensor/_ltensor.py:186]: cc: @albanD - do you want to comment on __torch_function__ vs. __torch_dispatch__

File: torch/distributed/tensor/_ltensor.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    return input_variant_axes - {axis_name}
+
+
+@register_variance_strategy([])  # [FIXME]: scatter ops
+def _variant_output_strategy(input_variant_axes, *args, **kwargs):
+    """Collectives that make output variant on the scatter axis (reduce_scatter)."""
+    return input_variant_axes
+
+
+class LTensor(torch.Tensor):
```
Comment (albanD) [inline torch/distributed/tensor/_ltensor.py:123]: If you want to disambiguate with the other local Tensor,
I would call this one a LDTensor for Local DTensor: the local representation of a DTensor.
I think this is the best way to go as I would expect this class will always be obtained by calling .to_local() on a DTensor?

File: torch/distributed/tensor/_ltensor.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+    """
+    ``LTensor`` (Local Tensor) is a subclass of ``torch.Tensor`` that tracks variance metadata
+    through DTensor -> Tensor -> DTensor transitions. It tracks mesh axes that a tensor varies along
+    (i.e., has different values across ranks)
+
+    States:
+    * Variant on axis: Tensor has different values across ranks on this mesh axis
+    * Invariant on axis: Tensor has identical values across ranks on this mesh axis (Replicate())
```
Comment (albanD) [inline torch/distributed/tensor/_ltensor.py]: Not sure what "(Replicate())" means here.

File: torch/distributed/tensor/_ltensor.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+                return t
+
+            local_tensor = t._local_tensor
+            src_variant_axes = t._variant_axes
+            missing_axes = out_variant_axes - src_variant_axes
+
+            if not missing_axes:
+                return local_tensor
+
+            from . import _varying_collectives as vcols
```
Comment (albanD) [inline torch/distributed/tensor/_ltensor.py]: import at the top

File: torch/distributed/tensor/_ltensor.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+    def __new__(
+        cls,
+        local_tensor: torch.Tensor,
+        variant_axes: set[str],
+        mesh: DeviceMesh,
+    ):
+        r = local_tensor.as_subclass(cls)
+        r._variant_axes = variant_axes
+        r._mesh = mesh
+        r._local_tensor = local_tensor
```
Comment (albanD) [inline torch/distributed/tensor/_ltensor.py:155]: isn't r the same as local_tensor here?

File: torch/distributed/tensor/_ltensor.py (hunk 4 above)
Comment (albanD) [reply torch/distributed/tensor/_ltensor.py:186]: We need to be able run new functions that are differentiable and show up in the autograd graph. So we have to be pre-autograd here.

File: torch/distributed/tensor/_varying_collectives.py (hunk 6 above)
Comment (vishal9-team) [reply torch/distributed/tensor/_varying_collectives.py:2]: I've added a design document at top with a section ([link](https://docs.google.com/document/d/1vi0UD-uaow3lZXWnnBlZIw_4011TS44HsSBS-bUkQwU/edit?tab=t.0#heading=h.4cu9dagbd0bu))

File: torch/distributed/tensor/_ltensor.py (hunk 8 above)
Comment (vishal9-team) [reply torch/distributed/tensor/_ltensor.py]: Tensor values are replicated along the ranks of the mesh axis

File: torch/distributed/tensor/_ltensor.py (hunk 10 above)
Comment (vishal9-team) [reply torch/distributed/tensor/_ltensor.py:155]: I'm holding this pointer for wrap/unwrap as __torch_function__ calls enter a recursive loop otherwise. I preferred this disabling torch function approach as it the latter seemed hacky

File: torch/distributed/tensor/experimental/_func_map.py (hunk 5 above)
Comment (ezyang) [inline torch/distributed/tensor/experimental/_func_map.py]: btw, another small nit, but PyTorch much more consistently uses "dim" instead of "axis" when talking about tensor dimensions. But also, here axis is referring to a mesh dim, not a tensor dim!
